core-tests/src/test/java/org/mule/runtime/core/privileged/transformer/CompositeConverterTestCase.java

```diff
@@ -167,4 +174,50 @@ public void appliesTransformerChainOnMessage() throws Exception {
     verify(transformationService, times(1)).applyTransformers(eq(testEvent().getMessage()), eq(testEvent()),
                                                               eq(compositeConverter));
   }
+
+  @Test
+  public void equalsReturnsTrueOnCompositeConvertersWithSameNameAndSameTransformationChain() {
+    Converter byteArrayToObjectConverter = new ByteArrayToObject();
+    Converter byteArrayToInputStreamConverter = new ByteArrayToInputStream();
+    CompositeConverter compositeConverterA = new CompositeConverter(byteArrayToObjectConverter, byteArrayToInputStreamConverter);
+    CompositeConverter compositeConverterB = new CompositeConverter(byteArrayToObjectConverter, byteArrayToInputStreamConverter);
+
+    assertThat(compositeConverterA, equalTo(compositeConverterB));
+  }
+
+
+  @Test
+  public void equalsReturnsFalseOnCompositeConvertersWithDifferentTransformationChain() {
+    Converter byteArrayToObjectConverter = new ByteArrayToObject();
+    Converter byteArrayToInputStreamConverter = new ByteArrayToInputStream();
+    CompositeConverter compositeConverterA = new CompositeConverter(byteArrayToObjectConverter, byteArrayToInputStreamConverter);
+    CompositeConverter compositeConverterB = new CompositeConverter(byteArrayToInputStreamConverter, byteArrayToObjectConverter);
+
+    assertThat(compositeConverterA, not(equalTo(compositeConverterB)));
+  }
+
+  @Test
+  public void hashCodeForCompositeConvertersChangesWithDifferentTransformationChain() {
+    Converter byteArrayToObjectConverter = new ByteArrayToObject();
+    Converter byteArrayToInputStreamConverter = new ByteArrayToInputStream();
+
+    Converter byteArrayToObjectConverter2 = new ByteArrayToObject();
+    Converter byteArrayToInputStreamConverter2 = new ByteArrayToInputStream();
+    Converter byteArrayToSerializableConverter = new ByteArrayToSerializable();
+
+    int hashCodeConverterA = new CompositeConverter(byteArrayToObjectConverter, byteArrayToInputStreamConverter).hashCode();
+
+    int hashCodeConverterAClone = new CompositeConverter(byteArrayToObjectConverter, byteArrayToInputStreamConverter).hashCode();
+
+    int hashCodeConverterAFakeClone =
+        new CompositeConverter(byteArrayToObjectConverter2, byteArrayToInputStreamConverter2).hashCode();
+    int hashCodeConverterReverseClone =
+        new CompositeConverter(byteArrayToInputStreamConverter, byteArrayToObjectConverter).hashCode();
+    int hashCodeConverterNotAClone =
+        new CompositeConverter(byteArrayToInputStreamConverter, byteArrayToSerializableConverter).hashCode();
+
+    assertThat(Arrays.asList(hashCodeConverterAClone), everyItem(equalTo(hashCodeConverterA)));
+    assertThat(Arrays.asList(hashCodeConverterReverseClone, hashCodeConverterNotAClone, hashCodeConverterAFakeClone),
```

Comment:
beware, hashCodes MUST be the same for equals object, but the opposite is not necessarily true.